File: client/whispervulkan.py
"""Reference client for the whispervulkan daemon (stdlib only).

    from whispervulkan import transcribe
    text = transcribe(open("clip.wav", "rb").read())
"""
import os
import json
import urllib.request

URL = os.environ.get("WHISPER_HTTP_URL", "http://127.0.0.1:48450/inference")


def transcribe(wav_bytes: bytes, response_format: str = "text", language: str = "") -> str:
    """POST 16 kHz mono WAV bytes to whispervulkan, return the transcript."""
    boundary = "----whispervulkan"
    parts = []
    fields = {"response_format": response_format}
    if language:
        fields["language"] = language
    for k, v in fields.items():
        parts.append(f"--{boundary}\r\nContent-Disposition: form-data; name=\"{k}\"\r\n\r\n{v}\r\n".encode())
    parts.append(
        f"--{boundary}\r\nContent-Disposition: form-data; name=\"file\"; "
        f"filename=\"audio.wav\"\r\nContent-Type: audio/wav\r\n\r\n".encode()
    )
    parts.append(wav_bytes)
    parts.append(f"\r\n--{boundary}--\r\n".encode())
    body = b"".join(parts)
    req = urllib.request.Request(
        URL, data=body,
        headers={"Content-Type": f"multipart/form-data; boundary={boundary}"},
    )
    with urllib.request.urlopen(req) as resp:
        raw = resp.read().decode("utf-8", "replace")
    if response_format == "json":
        return json.loads(raw).get("text", raw).strip()
    return raw.strip()


if __name__ == "__main__":
    import sys
    with open(sys.argv[1], "rb") as f:
        print(transcribe(f.read()))

File: client/whispervulkan.rs
//! Reference snippet for Tauri/Rust apps (e.g. Trik_Klip, voicechat).
//! Needs `reqwest` with the `multipart` + `blocking` features.
//!
//!   let text = whispervulkan_transcribe(&wav_bytes)?;
//!
//! WHISPER_HTTP_URL overrides the endpoint (default 127.0.0.1:48450).

pub fn whispervulkan_transcribe(wav: &[u8]) -> Result<String, Box<dyn std::error::Error>> {
    let url = std::env::var("WHISPER_HTTP_URL")
        .unwrap_or_else(|_| "http://127.0.0.1:48450/inference".to_string());
    let form = reqwest::blocking::multipart::Form::new()
        .text("response_format", "text")
        .part(
            "file",
            reqwest::blocking::multipart::Part::bytes(wav.to_vec())
                .file_name("audio.wav")
                .mime_str("audio/wav")?,
        );
    let text = reqwest::blocking::Client::new()
        .post(url)
        .multipart(form)
        .send()?
        .error_for_status()?
        .text()?;
    Ok(text.trim().to_string())
}
